apps/server/src/services/rundown-service/delayedRundown.utils.ts

```diff
@@ -171,6 +171,43 @@ export async function cachedReorder(eventId: string, from: number, to: number) {
   return reorderedEvent;
 }
 
+/**
+ * Swaps two events
+ * @param {string} fromEventId
+ * @param {string} toEventId
+ */
+export async function cachedSwap(fromEventId: string, toEventId: string) {
+  const fromEventIndex = DataProvider.getIndexOf(fromEventId);
+  const toEventIndex = DataProvider.getIndexOf(toEventId);
+
+  if (fromEventIndex === -1 || toEventIndex === -1) {
+    invalidateFromError();
+    throw new Error('ID not found at index');
+  }
+
+  const rundownToUpdate = DataProvider.getRundown();
+
+  const fromEvent = rundownToUpdate.at(fromEventIndex) as OntimeEvent;
+  const toEvent = rundownToUpdate.at(toEventIndex) as OntimeEvent;
+
+  rundownToUpdate[fromEventIndex] = {
+    ...toEvent,
+    timeStart: fromEvent.timeStart,
+    timeEnd: fromEvent.timeEnd,
+    duration: fromEvent.duration,
+  };
+
+  rundownToUpdate[toEventIndex] = {
+    ...fromEvent,
+    timeStart: toEvent.timeStart,
+    timeEnd: toEvent.timeEnd,
+    duration: toEvent.duration,
+  };
+
+  await DataProvider.setRundown(rundownToUpdate);
+  runtimeCacheStore.setCached(delayedRundownCacheKey, rundownToUpdate);
+}
+
```

Comment:
The caching abstraction could do with explaining. I will try my best to do it here:

The app uses a json object as the database, in that JSON object we store the raw rundown as the user enters.

From the persisted rundown (calling here user rundown), we extrapolate some more data and save the result in a new object, I will call this `delayedRundown`.
This serves two purposes:
- avoid keeping redundant / derived data in persistence (potential for bugs)
- normalise the data as we want to use in the app

In practice this means that we have
1. the user rundown, which we get with `DataProvider.getRundown()`
2. the delayedRundown, which we get from cache through the utility function `getDelayedRundown()`

When we mutate data in the rundown, we need to make sure that it happens in both the user rundown in `DataProvider` and the `delayedRundown`

When mutation we make a decision:
1. we are confident on what the change is, we make the mutate the object in the cached rundown and update it manually  `runtimeCacheStore.setCached(delayedRundownCacheKey, delayedRundown);`
2. we are unsure what needs to be recalculated, or we think that the recalculation might be more or as costly as calculating again, so we invalidate the whole thing `runtimeCacheStore.invalidate(delayedRundownCacheKey);`

Usually this would have been handled by a proxy. But since the `DataProvider` leverages a OOP architecture it is not possible to do so. The functions inside `delayedRundown.utils.ts` serve the same process. ie: not as convenient

As for your code, you have updated the user rundown, but not the `delayedRundown`.
You just need to mutate the rundown object as below (not great code, but hopefully explicit)

``` ts
export async function cachedSwap(fromEventId: string, toEventId: string) {
  const fromEventIndex = DataProvider.getIndexOf(fromEventId);
  const toEventIndex = DataProvider.getIndexOf(toEventId);

  if (fromEventIndex === -1 || toEventIndex === -1) {
    invalidateFromError();
    throw new Error('ID not found at index');
  }

  const rundownToUpdate = DataProvider.getRundown();

  const fromEvent = rundownToUpdate.at(fromEventIndex) as OntimeEvent;
  const toEvent = rundownToUpdate.at(toEventIndex) as OntimeEvent;

  rundownToUpdate[fromEventIndex] = {
    ...toEvent,
    timeStart: fromEvent.timeStart,
    timeEnd: fromEvent.timeEnd,
    duration: fromEvent.duration,
  };

  rundownToUpdate[toEventIndex] = {
    ...fromEvent,
    timeStart: toEvent.timeStart,
    timeEnd: toEvent.timeEnd,
    duration: toEvent.duration,
  };


  // ADDED CODE STARTS HERE
  const delayedRundown = getDelayedRundown();

  const fromCachedEvent = delayedRundown.at(fromEventIndex) as OntimeEvent;
  const toCachedEvent = delayedRundown.at(toEventIndex) as OntimeEvent;

  if (fromCachedEvent.id !== fromEvent.id || toCachedEvent.id !== toEvent.id) {
    // something went wrong, we invalidate the cache
    runtimeCacheStore.invalidate(delayedRundownCacheKey);
  } else {
    delayedRundown[fromEventIndex] = {
      ...toCachedEvent,
      timeStart: fromCachedEvent.timeStart,
      timeEnd: fromCachedEvent.timeEnd,
      duration: fromCachedEvent.duration,
    };

    delayedRundown[toEventIndex] = {
      ...fromCachedEvent,
      timeStart: toCachedEvent.timeStart,
      timeEnd: toCachedEvent.timeEnd,
      duration: toCachedEvent.duration,
    };
    runtimeCacheStore.setCached(delayedRundownCacheKey, delayedRundown);
  }

  // ADDED CODE FINISHES HERE

  await DataProvider.setRundown(rundownToUpdate);

```

Reply:
Awesome. Thank you for the lengthy reply!

Resolved with https://github.com/cpvalente/ontime/pull/446/commits/3b4ec45c12a705423fcb5459d267a9bc907b8598